src/data.rs

```diff
@@ -0,0 +1,137 @@
+use super::utils::collections::{btree_map, BTreeMap, BTreeSet};
+use core::{cell::RefCell, iter::Chain};
+
+/// A [DataRecorder] that records read requests to the underlying key-value map.
+/// The data recorder is used to generate a proof for read requests. The proof is used to
+/// enable stateless execution of programs on the Miden virtual machine.
+///
+/// The [DataRecorder] is composed of three parts:
+/// - `init`: which contains the initial key-value pairs from the underlying data set.
+/// - `new`: which contains key-value pairs which are created during transaction execution.
+/// - `recorder`: which contains the keys from the initial data set that are read during
+///   program execution.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct DataRecorder<K, V> {
+    init: BTreeMap<K, V>,
+    new: BTreeMap<K, V>,
+    recorder: RefCell<BTreeSet<K>>,
+}
+
+impl<K: Ord + Clone, V: Clone> DataRecorder<K, V> {
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+    /// Returns a new [DataRecorder] instance initialized with the provided key-value pairs
+    /// ([BTreeMap]).
+    pub fn new(init: BTreeMap<K, V>) -> Self {
+        DataRecorder {
+            init,
+            new: BTreeMap::new(),
+            recorder: RefCell::new(BTreeSet::new()),
+        }
+    }
+
+    /// ACCESSORS
+    /// --------------------------------------------------------------------------------------------
+    /// Returns a reference to the value associated with the given key if the value exists. If the
+    /// key is part of the initial data set, the key access is recorded.
+    pub fn get(&self, key: &K) -> Option<&V> {
+        if let Some(value) = self.new.get(key) {
+            return Some(value);
+        }
+
+        match self.init.get(key) {
+            None => None,
+            Some(value) => {
+                self.recorder.borrow_mut().insert(key.clone());
+                Some(value)
+            }
+        }
+    }
+
+    /// Returns a boolean to indicate whether the given key exists in the data set. If the key is
+    /// part of the initial data set, the key access is recorded.
+    pub fn contains_key(&self, key: &K) -> bool {
+        if self.new.contains_key(key) == true {
+            return true;
+        }
+
+        match self.init.contains_key(key) {
+            true => {
+                self.recorder.borrow_mut().insert(key.clone());
+                true
+            }
+            false => false,
+        }
+    }
+
+    // TODO: Consider if we are happy with the fact that duplicates are counted twice in both
+    // `len(..)` and `iter(..)` methods.
```

Comment:
If we are unhappy with double counting then we can introduce a counter for duplicates which we deduct when `len` is called.  To make the `iter(..)` method only return unique values we could use `itertools` or alternatively copy the [`Unique` implementation](https://docs.rs/itertools/latest/itertools/trait.Itertools.html#method.unique) we require .

Reply:
I think we do want to handle this properly (especially for iteration) as it may affect how things work downstream. Depending on how much work it is, copying the implementation would be better (though, if it proves rather complex, adding `itertools` as a dependency may be OK.

Another potential alternative is to change how this data structure keeps track of updates. Specifically, it could look something like this:
```Rust
pub struct DataRecorder<K, V> {
    data: BTreeMap<K, V>,
    updates: BTreeSet<K>,
    trace: RefCell<BTreeMap<K, V>>,
}
```
And something like `get()` method could then work like so:
```Rust
pub fn get(&self, key: &K) -> Option<&V> {
    match self.state.get(key) {
        None => None,
        Some(value) => {
            if !self.updates.contains(key) {
                self.trace.borrow_mut().insert(key.clone(), value);                
            }
            Some(value)
        }
    }
}
```
This does do more work - so, not sure if it is worth it.

Reply:
What does state represent in this instance?

Reply:
oh - sorry, `state` should have been `data`.

Reply:
I don't think your proposal will work. It will always return the original value even if it has been updated. I implemented an alternate scheme, this involves the introduction of `updated_keys` which keeps track of the keys from the original set which have been updated.

Reply:
I've implemented this approach in 80a7f019161866c18a6e0db46b370e2d81b77cc4 as it simplified the implementation a bit. The tests seem to be passing (and I've added some more checks) - but take a look and see if I missed anything.